nix/shell.nix

```diff
@@ -0,0 +1,126 @@
+{ ghc, pkgs, lib, project, pre-commit-hooks }:
+
+let
+
+  toolVersions = {
+    cabal = "latest";
+    cabal-fmt = "latest";
+    fourmolu = "latest";
+    haskell-language-server = "latest";
+    hlint = "latest";
+    stylish-haskell = "latest";
+  };
+
+  toolVariants =
+    lib.attrsets.mapAttrs
+      (_: ghcVariant: lib.attrsets.mapAttrs ghcVariant.tool toolVersions)
+      project.projectVariants;
+
+  toolOverrides = {
+    ghc910.cabal-fmt = toolVariants.ghc96.cabal-fmt; # cabal-fmt not buildable with ghc9102
+    ghc912.cabal-fmt = toolVariants.ghc96.cabal-fmt; # cabal-fmt not buildable with ghc9122
+  };
+
+  tools = toolVariants.${ghc} // (toolOverrides.${ghc} or { });
+
+  preCommitCheck = pre-commit-hooks.lib.${pkgs.system}.run {
+
+    src = lib.cleanSources ../.;
+
+    hooks = {
+      cabal-fmt = {
+        enable = false;
+        package = tools.cabal-fmt;
+      };
+      fourmolu = {
+        enable = true;
+        package = tools.fourmolu;
+      };
+      hlint = {
+        enable = true;
+        package = tools.hlint;
+      };
+      stylish-haskell = {
+        enable = false;
+        package = tools.stylish-haskell;
+      };
```

Comment:
Why do we need stylish haskell?

Reply:
We don't, so I left it disabled. However, since it's one of the tools provided by haskell.nix and supported by pre-commit, I left an entry for it as a placeholder in case anyone ever wanted it in the future. That seems unlikely, though, since fourmolu supersedes it, so I'll take it out.

I also noticed that I accidentally left `cabal-fmt` disabled, so I'll enable it now.